docs/misc/str-builder-api.rst

```diff
@@ -0,0 +1,388 @@
+bytes/str building API
+=======================
+
+We need to design an HPy API to build ``bytes`` and ``str`` objects. Before making
+any proposal, it is useful to understand:
+
+1. What is the current API to build strings.
+
+2. What are the constraints for alternative implementations and the problems
+   of the current C API.
+
+3. What are the patterns used to build string objects in the existing
+   extensions.
+
+Some terminology:
+
+    * "string" indicates either ``bytes`` or ``str`` objects aaa
+    * "unicode" or "unicode string" indicates ``str``
+
+.. note::
+   In this issue we are translating ``PyUnicode_*`` functions into
+   ``HPyStr_*``. See issue #xxx for more discussion about the naming convention.
+
+Current CPython API
+--------------------
+
+Bytes
+~~~~~
+
+There are essentially two ways to build ``bytes``:
+
+1. Copy the content from an existing C buffer:
+
+.. code-block:: c
+
+    PyObject* PyBytes_FromString(const char *v);
+    PyObject* PyBytes_FromStringAndSize(const char *v, Py_ssize_t len);
+    PyObject* PyBytes_FromFormat(const char *format, ...);
+
+
+2. Create an uninitialized buffer and fill it manually:
+
+.. code-block:: c
+
+    PyObject s = PyBytes_FromStringAndSize(NULL, size);
+    char *buf = PyBytes_AS_STRING(s);
+    strcpy(buf, "hello");
+
+(1) is easy for alternative implementations and we can probably provide an HPy
+equivalent without changing much, so we will concentrate on (2): let's call it
+"raw-buffer API".
+
+Unicode
+~~~~~~~
+
+Similarly to ``bytes``, there are several ways to build a ``str``:
+
+.. code-block:: c
+
+    PyObject *PyUnicode_FromString(const char *u);
+    PyObject* PyUnicode_FromStringAndSize(const char *u, Py_ssize_t size);
+    PyObject* PyUnicode_FromKindAndData(int kind, const void *buffer, Py_ssize_t size);
+    PyObject* PyUnicode_FromFormat(const char *format, ...);
+    PyObject* PyUnicode_New(Py_ssize_t size, Py_UCS4 maxchar);
+
+
+.. note::
+   ``PyUnicode_FromString{,AndSize}`` take an UTF-8 string in input
+
+The following functions are used to initialize an uninitialed object, but I
+could not find any usage of them outside CPython itself, so I think they can
+be safely ignored for now:
```

Comment:
As another point of reference: PyPy does not implement them, and no issue has been reported that they are missing.